### ISSUE TITLE
Duet in docker container

### ISSUE THREAD

**Issue by gmuraru** [OpenMined/SyMPC#187]
## Feature Description
Currently, we run duet manually by opening 2 Jupiter notebooks - one for DS and one for DO.
We would want to have 3 docker containers:
1. Signaling server
2. Data Scientist
3. Data Owner

* The DS and DO notebooks would allow you to connect to Jupyter notebook server (one for the DS and one for DO) - this for more interactive stuff.
* We also want to have the possibility to run scripts directly in the docker container - by providing a python script.

## Is your feature request related to a problem?
Have an easier way to test Duet Notebooks.

## Additional Context
You need to check how to run Duet notebooks in the [Syft repository](https://github.com/OpenMined/PySyft/tree/dev/packages/syft).

**Comment by gmuraru** [comment]
@madhavajay @tudorcebere pinging you here because this might be something that we implement in Syft?

**Comment by danielorihuela** [comment]
If there is no problem, I am going to work on this. I understand the Data Scientist (person working on data) and the Data Owner (decides which data to share to whom). But I do not understand what is the Signaling Server. Could you explain it? Thanks!

**Comment by danielorihuela** [comment]
Another question. I guess that the jupyter notebooks of the owner and the scientist, that will be in a docker, must be accessible through a browser. As if the user installed and started running it, right?

**Comment by gmuraru** [comment]
> If there is no problem, I am going to work on this. I understand the Data Scientist (person working on data) and the Data Owner (decides which data to share to whom). But I do not understand what is the Signaling Server. Could you explain it? Thanks!

From what I know the "signaling server" is a component (it is in Syft cc: @tudorcebere ) that makes the DS and DO know about each other.

Yeah, ideally they would have have a jupyter notebook server and we can connect to them using different addresses

**Comment by danielorihuela** [comment]
Until now I have a docker compose that starts two docker containers. The data owner one (port 8888) and the scientist one (port 8889). I have tried to use duet from two different computers inside the same network and it does work without using the signaling server. So we do not need it @gmuraru. By default, syft uses the public signaling server or Openmined.

For the moment I will continue with the two following points:
1. Copy the examples to the containers. That way it may be easier for new comers to try the different things (if they wish to use docker).
2. Work on automatically pairing it (something that @gmuraru commenetd me in a PM the other day).

The signaling server could be useful to local tests, that way a person could try the examples locally without internet. That could be interesting.

**Comment by gmuraru** [comment]
I think you can also use ```loopback=True``` in the ```Duet``` class (when you create the duet class).
By using that, I think we might "get rid" of the need to implement the signaling server?

**Comment by danielorihuela** [comment]
No no. It already works!! We do not need a docker container for the signaling server. By default it uses the public one in amazon. I just wanted to add that we could do the docker container for a full local, without the need of using the Openmined public one. I do not know if you did not understand me or I am not understanding you.

**Comment by danielorihuela** [comment]
I have been doing some research and I did not find a way to to the autopair of duet notebooks through the Openmined public signaling server. Furthermore, I did not find any way to execute jupyter notebooks automatically.
So, I will upload what I have done until now. Docker compose that start two docker containers, one for the owner and one for the scientist. Each one with his own notebook (only the commands to launch the duet server and join the duet).
